core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/CompactExpressionEvaluator.scala

```diff
@@ -135,179 +140,213 @@ class CompactExpressionEvaluator(repository: CompactEntityRepository) extends Ex
-
-                val rootEntityNames = if (entityType == rootEntityType) {
-                  Seq(entityName)
-                } else {
-                  combinedExpressionAndResults.flatMap { case (_, resultMap) => resultMap.keys }.distinct
-                }
-
-                // Pre-compute a map from expression -> ExpressionAndResult for O(1) lookup
-                val entityExpressionToResultMap: Map[String, Seq[ExpressionAndResult]] =
-                  combinedExpressionAndResults.groupBy { case (expression, _) => expression }
-
-                // Transform workspaceExpressionResults to the same shape and add to the map
-                val workspaceExpressionAndResults: Seq[ExpressionAndResult] = workspaceExpressionResults.map {
-                  case (lookupExpression, tryValues) =>
-                    // Create a result map with the workspace values for all root entity names
-                    val resultMap = rootEntityNames.map { entityName =>
-                      entityName -> tryValues
-                    }.toMap
 
-                    (lookupExpression, resultMap)
-                }.toSeq
-
-                val workspaceExpressionToResultMap: Map[String, Seq[ExpressionAndResult]] =
-                  workspaceExpressionAndResults.groupBy { case (expression, _) => expression }
-
-                // Combine both maps, with workspace results taking precedence if there are conflicts
-                val expressionToResultMap: Map[String, Seq[ExpressionAndResult]] =
-                  entityExpressionToResultMap ++ workspaceExpressionToResultMap
+              // Pre-compute a map from expression -> ExpressionAndResult for O(1) lookup
+              val entityExpressionToResultMap: Map[String, Seq[ExpressionAndResult]] =
+                combinedExpressionAndResults.groupBy { case (expression, _) => expression }
+
+              // Transform workspaceExpressionResults to the same shape and add to the map
+              val workspaceExpressionAndResults: Seq[ExpressionAndResult] = workspaceExpressionResults.map {
+                case (lookupExpression, tryValues) =>
+                  // Create a result map with the workspace values for all root entity names
+                  val resultMap = rootEntityNames.map { entityName =>
+                    entityName -> tryValues
+                  }.toMap
+
+                  (lookupExpression, resultMap)
+              }.toSeq
+
+              val workspaceExpressionToResultMap: Map[String, Seq[ExpressionAndResult]] =
+                workspaceExpressionAndResults.groupBy { case (expression, _) => expression }
+
+              // Combine both maps, with workspace results taking precedence if there are conflicts
+              val expressionToResultMap: Map[String, Seq[ExpressionAndResult]] =
+                entityExpressionToResultMap ++ workspaceExpressionToResultMap
+
+              if (inputExpressionData.isEmpty) {
+                // Short-circuit if there are no inputs
```

Comment:
This is the main change in this method - check for lack of inputs and return results, mapped to the entities.  Unfortunately, in order to account for returning different types at different parts of the method now, I've done quite a bit of rearrangement, even though the actual substance of the method hasn't otherwise changed.